### PR TITLE
WIP: search multiple branches if indexed

### PR DESCRIPTION
**Pull request by keegancsmith**

### CODE REVIEW

**Comment by keegancsmith** [comment]
@rvantonder an early review on `zoekt: handle multiple branches in zoektIndexedRepos` would be appreciated since you wrote the original logic in `zoektSingleIndexedRepos`. Note the commit message for it, this implementation will allow us to remove that special casing :)

File: cmd/frontend/graphqlbackend/zoekt.go [hunk 1]
```diff
@@ -456,96 +456,67 @@ func queryToZoektFileOnlyQueries(query *search.TextPatternInfo, listOfFilePaths
...
 		}
 
-		repo, ok := set[strings.ToLower(string(rev.Repo.Name))]
-		if !ok || (filter != nil && !filter(repo)) {
-			unindexed = append(unindexed, rev)
+		revspecs := reporev.RevSpecs()
+
+		if len(revspecs) != len(reporev.Revs) {
+			// Contains a RefGlob or ExcludeRefGlob so we can't do indexed
+			// search on it.
```
Comment (rvantonder) [inline cmd/frontend/graphqlbackend/zoekt.go:486]: A minor comment, I guess--when reading this comment I'm wondering if this PR or another could include some more naming that's more indicative of what is incoming requests versus data we have like

"requestedRevs" or something for incoming requests and "resolvedRevs" or similar that we're checking against.

**Comment by rvantonder** [comment]
LGTM so far :-)

File: cmd/frontend/graphqlbackend/zoekt.go [hunk 2]
```diff
@@ -456,96 +456,67 @@ func queryToZoektFileOnlyQueries(query *search.TextPatternInfo, listOfFilePaths
...
+	unindexed = make([]*search.RepositoryRevisions, 0)
 
-	for _, rev := range revs {
-		if len(rev.RevSpecs()) >= 2 || len(rev.RevSpecs()) != len(rev.Revs) {
-			// Zoekt only indexes 1 rev per repository, so it will not have the full results for the
-			// query on repositories for which multiple revs are searched.
-			unindexed = append(unindexed, rev)
+	for _, reporev := range revs {
+		repo, ok := set[strings.ToLower(string(reporev.Repo.Name))]
+		if !ok || (filter != nil && !filter(repo)) {
```
Comment (tsenart) [inline cmd/frontend/graphqlbackend/zoekt.go:477]: Can we call this `isIndexed` rather than `filter` which is always ambiguous if it's filtering in or out?

Comment (keegancsmith) [reply cmd/frontend/graphqlbackend/zoekt.go:477]: yup. Note this is cut pasted code not my own creation :P Also I'll do cosmetic changes like this at a later stage, this code is still in flux. Additionally this was added purely for symbol search, I'll see if we can do something more explicit than an opaque filter function that is only set by symbol search.

**Comment by keegancsmith** [comment]
A quick note of the current state. I've been hacking on zoektIndexedRepo returning the RepoBranches to search. This works really well, except breaks down for one requirement. Currently we map the search "input rev" to the filematch.uri. In just HEAD world we always used revs[0]. In multi-branch world we lose this mapping since zoekt just returns which branches matched. I have some ideas I have been trying, but this bit of bookkeeping somewhat introduces a very tight coupling between zoektIndexedRepos and zoekSearchHEAD. The next idea I am experimenting with is not mapping input rev, but picking either the branch or SHA returned by zoekt based on some heuristics. This should retain the expected user behaviour (and may even be more correct).

**Comment by keegancsmith** [comment]
This branch ended up being split out and re-worked. So this feature has landed so can close this.

**Pull request closed without merging**